### PR TITLE
Restore dosen't work #1016

### PR DESCRIPTION
**Pull request by smoriarty21**
Fix configuration load/save for board and sensor alignment, craft name, DShot Beacon, ESC/Motor features, beeper, mixer, accelerometer, barometer and magnetometer

### CODE REVIEW

File: src/js/backup_restore.js [hunk 1]
```diff
@@ -309,8 +317,8 @@ function configuration_restore(callback) {
                             GUI.log(i18n.getMessage('backupFileUnmigratable'));
                             return;
                         }
-
-                        if (configuration.FEATURE_CONFIG.features._featureMask) {
+
```
Comment (mikeller) [inline src/js/backup_restore.js]: Extra whitespace.

**Comment by mikeller** [comment]
Ok, this seems to be not quite ready. I have pulled the attached backup from a functioning flight controller. After a restore, the flight controller will reproducibly wedge until it is forced into DFU mode and flashed with 'Full chip erase':


[BTFL_backup_x_20180626_222343.json.zip](https://github.com/betaflight/betaflight-configurator/files/2136808/BTFL_backup_x_20180626_222343.json.zip)

**Comment by smoriarty21** [comment]
@mikeller Thanks I'll take a look and fix it up

**Comment by smoriarty21** [comment]
@mikeller Give this one a try, works on my SPRacing F3 now with the given backup.  I do have an issue with the PID loop timing as it seems the default isn't valid for my board.  I wanted to double check with you that is just my old F3 and not an issue with the default value, whats your thought on that?  I'll jump on the issue with the config not pulling on connect and that should make all these problems go away but, I wanted to make sure to at least run it by you

File: src/js/msp/MSPHelper.js [hunk 2]
```diff
@@ -1222,7 +1224,7 @@ MspHelper.prototype.crunch = function(code) {
             if (semver.gte(CONFIG.apiVersion, "1.37.0")) {
                 buffer.push8(BEEPER_CONFIG.dshotBeaconTone);
             }
-            if (semver.gte(CONFIG.apiVersion, "1.39.0")) {
+            if (semver.gte(CONFIG.apiVersion, "1.39.0") && BEEPER_CONFIG.dshotBeaconConditions.getMask() != 0) {
```
Comment (mikeller) [inline src/js/msp/MSPHelper.js]: This won't work. API 1.39 expects this value to be sent, and if it's not sent the firmware will read whatever is sent after it and use it. Also, a mask of `0` is valid.

File: src/js/msp/MSPHelper.js [hunk 3]
```diff
@@ -1458,6 +1460,9 @@ MspHelper.prototype.crunch = function(code) {
                 .push8(SENSOR_ALIGNMENT.align_mag);
             break;
         case MSPCodes.MSP_SET_ADVANCED_CONFIG:
+            if (PID_ADVANCED_CONFIG.motor_pwm_rate == 0 || (semver.gte(CONFIG.apiVersion, "1.24.0") && PID_ADVANCED_CONFIG.digitalIdlePercent == 0)) {
```
Comment (mikeller) [inline src/js/msp/MSPHelper.js]: Again, this won't work, as the sender of `MSP_SET_ADVANCED_CONFIG` expects this to be sent irrespective of values.

File: src/js/backup_restore.js [hunk 4]
```diff
@@ -165,13 +170,72 @@ function configuration_backup(callback) {
...
 
                 save();
             }
         }
+
+        if (GUI.configuration_loaded === true) {
+            return fetch_unique_data_item();
+        }
+        var promises = [ getAdvancedConfig(), getSensorConfig(), getCraftName(), getBoardAlignmentConfig(), getMixerConfig(), getBeeperConfig() ];
+        Promise.all(promises).
```
Comment (mikeller) [inline src/js/backup_restore.js]: This is a risky way of doing things. `Promise.all` does not specify if the promises are resolved in sequence or in parallel. Even if it works for now, the implementation might change (or might be different on a different platform), resulting in the MSP calls being executed in parallel.
These should be sequentialised like in:

https://github.com/smoriarty21/betaflight-configurator/blob/70c839eb3a3f869672af15992e883774d5497a00/src/js/tabs/pid_tuning.js#L23-L43

**Pull request merged**